document-portal/document-portal.c

```diff
@@ -1416,6 +1416,90 @@ portal_list (GDBusMethodInvocation *invocation,
   return TRUE;
 }
 
+const char *
+get_host_path_internal (GDBusMethodInvocation *invocation,
+             XdpAppInfo *app_info,
+             const char *id)
+{
+  g_autoptr(PermissionDbEntry) entry = NULL;
+
+  XDP_AUTOLOCK (db);
+
+  entry = permission_db_lookup (db, id);
+
+  if (!entry)
+    {
+      g_dbus_method_invocation_return_error (invocation,
+                                             XDG_DESKTOP_PORTAL_ERROR, XDG_DESKTOP_PORTAL_ERROR_INVALID_ARGUMENT,
+                                             g_strdup_printf("Invalid ID passed (%s)", id));
+      return NULL;
+    }
+
+  if (!xdp_app_info_is_host (app_info))
+    {
+      g_autofree const char **apps = NULL;
+      const char *app_id = NULL;
+      gboolean app_found = FALSE;
+      int i;
+
+      app_id = xdp_app_info_get_id(app_info);
+
+      apps = permission_db_entry_list_apps (entry);
+      for (i = 0; apps[i] != NULL; i++)
+        {
+          if (g_strcmp0 (app_id, apps[i]) == 0)
+            {
+              app_found = TRUE;
+              break;
+            }
+        }
+
+      if (!app_found)
+        {
+          g_dbus_method_invocation_return_error (invocation,
+                                                 XDG_DESKTOP_PORTAL_ERROR, XDG_DESKTOP_PORTAL_ERROR_NOT_ALLOWED,
+                                                 "Not enough permissions");
+          return NULL;
+        }
+    }
+
+  g_autoptr (GVariant) data = permission_db_entry_get_data (entry);
+  const char *path;
+
+  g_variant_get (data, "(^&ayttu)", &path, NULL, NULL, NULL);
+
+  return path;
```

Comment:
This seems wrong to me: (https://docs.gtk.org/glib/gvariant-format-strings.html#convenience-conversions)

> ^&ay 		equivalent to g_variant_get_bytestring()

and for `g_variant_get_bytestring`:

> The return value remains valid as long as value exists.

I.e, path becomes invalid at the end of the function because data will be freed.

(also declarations should be at the start of the block)

```suggestion
  data = permission_db_entry_get_data (entry);

  g_variant_get (data, "(^ayttu)", &path, NULL, NULL, NULL);

  return path;
```

but this needs to be freed in `portal_get_host_paths` then.